torch/_inductor/autotune_process.py

```diff
@@ -150,15 +148,130 @@ def close(self) -> None:
             self.process.wait()
             self.close()
 
 
-tuning_process = TuningProcess()
+@dataclasses.dataclass
+class TuningProcessPool:
+    """
+    Maintains a pool of TuningProcesses to benchmark kernels in parallel
+    across devices. By default, we create one TuningProcess per device and
+    set the sub-process environment to make only that device visible.
+    """
+
+    processes: Optional[Queue[TuningProcess]] = None
+    executor: Optional[ThreadPoolExecutor] = None
+
+    def initialize(self, count: Optional[int] = None) -> None:
+        """
+        Start the child processes.
+        """
+        assert (self.processes is None) == (self.executor is None)
+        if self.processes is not None:
+            return
+
+        if config.autotune_multi_device:
+            count = count or torch.cuda.device_count()
+        else:
+            count = 1
+        assert count > 0 and count <= torch.cuda.device_count()
+
+        # Launch the child processes and push a msg to "warm up"
+        self.processes = Queue()
+        for device in range(count):
+            p = TuningProcess(device=device if config.autotune_multi_device else None)
+            p.initialize()
+            p.put(Ping())
+            self.processes.put(p)
+
+        # Wait for the initialization to finish
+        for p in self.processes.queue:
+            assert isinstance(p.get(), Pong)
+
+        # Use a thread pool to manage distributing work to the subprocesses.
+        # Threads block on an available process, so it makes sense to match
+        # the number of threads with the number of devices.
+        self.executor = ThreadPoolExecutor(max_workers=count)
+
+        # Register the exit handler for the parent process so it will terminate
+        # the child processes.
+        global EXIT_HANDLER_REGISTERED
+        if not EXIT_HANDLER_REGISTERED:
+            EXIT_HANDLER_REGISTERED = True
+            import atexit
+
+            atexit.register(lambda: self.terminate())
+
+    def terminate(self) -> None:
+        """
+        Signal all child processes to terminate.
+        """
+        if self.executor is not None:
+            self.executor.shutdown()
+            self.executor = None
+
+        if self.processes is not None:
+            for p in self.processes.queue:
+                p.terminate()
+            for p in self.processes.queue:
+                p.wait()
+            self.processes = None
+
+    def target(self, choice: TritonTemplateCaller) -> float:
+        """
+        Entry point for the thread-pool helper threads: Wait for an open TuningProcess,
+        remove it from the queue, execute the benchmark in that subprocess, and return
+        the TuningProcess to the queue.
+        """
+        assert choice.bmreq is not None
+        assert self.processes is not None
+
+        process = self.processes.get()
+        process.put(choice.bmreq)
+        try:
+            return process.get()
```

Comment:
Should there be a timeout? Can this get stuck forever otherwise?

Reply:
@aakhundov, I'm not sure. Do you have a scenario in mind that would lead to a sub-process hanging? I'd love a repro to play around with. I'm new here and my naive assumption was that these kernels are fairly constrained and we wouldn't be concerned about arbitrary code. If the process crashes or exits unexpectedly, then we handle that scenario.

Reply:
I don't have an exact scenario in mind. Just, given that we're relying on STDIN- / STDOUT-based communication, assuming that something can go wrong there leading to indefinitely long waiting. In AIT, we have a timeout for the whole profiling session (which, to be fair, we had to increase a couple of times due to longer e2e profiling that was initially expected).